core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala

```diff
@@ -0,0 +1,772 @@
+    doPrint("Verify index create and drop for various index types")
+
+    snContext.sql(s"create index $index1 on $table1 " +
+        s" (COL2, COL3) Options (colocate_with  '$table2')")
+    indexesToDrop += index1
+
+    snContext.sql(s"create index $index2 on $table1 " +
+        s" (COL1, COL3) Options (colocate_with  '$table3')")
+    indexesToDrop += index2
+
+    snContext.sql(s"create index $index3 on $table1 " +
+        s" (COL4) Options (colocate_with  '$table4')")
+    indexesToDrop += index3
+
+    val executeQ = QueryExecutor(snContext, false, false)
+    /*
+    executeQ(s"select t1.col2, t2.col3 from $table1 t1, $table2 t2, $table4 t4 " +
+        s"where t1.col2 = t2.col2 and t1.col3 = t2.col3 " +
+        s"and t1.col4 = t4.col4 and t1.col1 = t4.col2 ")
+*/
+
+    executeQ(s"select t1.col2, t2.col3 from $table1 t1, $table2 t2 " +
+        s"where t1.col2 = t2.col2 and t1.col3 = t2.col3 ") {
+      validateIndex(Seq(index1), table2)(_)
+    }
+
+    executeQ(s"select t1.col2, t2.col3 from $table1 t1, $table2 t2, $table3 t3 " +
+        s"where t1.col2 = t2.col2 and t1.col3 = t2.col3 " +
+        s"and t1.col1 = t3.col1 and t1.col3 = t3.col3") {
+      // t1 -> t2, t1 -> t3
+      validateIndex(Seq(index1), table2, table3)(_)
+    }
+
+    executeQ(s"select t1.col2, t4.col3 from $table1 t1, $table4 t4 " +
+        s"where t1.col4 = t4.col4 and t1.col1 = t4.col2 ") {
+      // still results into shuffle.
+      validateIndex(Seq(index3), table4)(_)
+    }
+
+    executeQ(s"select t1.col2, t2.col3 from $table1 t1, $table2 t2, $table4 t4 " +
+        s"where t1.col2 = t2.col2 and t1.col3 = t2.col3 " +
+        s"and t1.col4 = t4.col4 ") {
+      // t1 -> t2, t1 -> t4
+      validateIndex(Seq(index1), table2, table4)(_)
+    }
+
+    executeQ(s"select t1.col2, t2.col3 from $table1 t1, $table4 t4, $table2 t2 " +
+        s"where t1.col2 = t2.col2 and t1.col3 = t2.col3 " +
+        s"and t1.col4 = t4.col4 ") {
+      // t1 -> t4, t1 -> t2
+      validateIndex(Seq(index1), table2, table4)(_)
+    }
+
+    executeQ(s"select t1.col4, xx.col5 from $table1 t1, " +
+        s"(select t4.col1 as col5 from $table4 t4) xx, " +
+        s"$table2 t2 where t1.col2 = t2.col2 and t1.col3 = t2.col3 " +
+        s"and t1.col4 = xx.col5 ") {
+      // t1 -> t4, t1 -> t2
+      validateIndex(Seq(index1), table2, table4)(_)
+    }
+
+    executeQ(s"select t1.col4, xx.col5 from $table1 t1, " +
+        s"(select t4.col1 as col5, col2, col3 from $table4 t4) xx, " +
+        s"$table2 t2 where xx.col2 = t2.col2 and xx.col3 = t2.col3 " +
+        s"and t1.col4 = xx.col5 ") {
+      // t1 -> t4, t2 -> t4
+      validateIndex(Seq.empty, table1, table2, table4)(_)
+    }
+
+    executeQ(s"select t1.col2, t2.col3 from $table1 t1, $table4 t4, $rtable5 t5, $table2 t2 " +
+        s"where t1.col4 = t4.col4 " +
+        s"and t1.col2 = t5.col2 and t1.col3 = t5.col3 " +
+        s"and t5.col2 = t2.col2 and t5.col3 = t2.col3 ") {
+      // t1 -> t4, t1 -> t5 -> t2
```

Comment:
@soubhik-c  Doesn't this require queryHint "includeGeneratedPaths" or it is enabled by default ?